datafusion/expr/src/logical_plan/tree_node.rs

```diff
@@ -81,7 +79,7 @@ impl TreeNode for LogicalPlan {
                 expr,
                 input,
                 schema,
-            }) => rewrite_arc(input, f)?.update_data(|input| {
+            }) => input.map_elements(f)?.update_data(|input| {
```

Comment:
Is the difference between `map_children` and `map_elements` that `map_elements` doesn't also apply the function to `input` ?

Reply:
`map_children` also does not apply the f to input. IIUC `map_children` defines how LogicalPlan's construct a tree topology, while `map_elements` is resolving the link between nodes.

Reply:
> IIUC map_children defines how LogicalPlan's construct a tree topology, while map_elements is resolving the link between nodes.

Yes, the container trait supposes that the elements in it are sibling nodes. Basically this is why we can use it when we implement `apply_children()` / `map_children()`. The blanket implementations for the trait makes it very simple to use with `Expr` and `LogicalPlan` as those logical plan treenodes define their children as a composition of `Vec`s, `Option`s, tuples, `Box`es and `Arc`s.

Here in `LogicalPlan` a treenode has mostly only 1 child (except for `Join` and `Union` maybe) so input (`Arc<LogicalPlan>`) is just a single element container so this PR doesn't have much benefit, but I think it is worth to do it to allign `Expr` and `LogicalPlan`.

`TreeNode::map_children` and `TreeNodeContainer:: map_elements` are similar in a way that both apply `f` to the children/elements of a `TreeNode` / `TreeNodeContainer`. But we use `TreeNode` to define different kids of logical 
 and physical plan trees, while `TreeNodeContainer` can be used to simplify implementation of `apply_children` / `map_children()`.

Reply:
I clarified this in https://github.com/apache/datafusion/pull/13467/commits/2097bc7603097f4106d68be9f1b6a281d4bdbb7e.